pandas/core/frame.py

```diff
@@ -1337,6 +1344,10 @@ def to_numpy(self, dtype=None, copy: bool = False) -> np.ndarray:
                [2, 4.5, Timestamp('2000-01-02 00:00:00')]], dtype=object)
         """
         result = np.array(self.values, dtype=dtype, copy=copy)
```

Comment:
For some dtypes, it will not work to first convert to an array of that dtype, and only afterwards to replace missing values. 
For example, with Int column with missing value, the initial step to convert to a float array already fails:

```
In [22]: pd.DataFrame({'a': [1, 2, None]}, dtype="Int64").to_numpy(float) 
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-22-2847bf76b2be> in <module>
----> 1 pd.DataFrame({'a': [1, 2, None]}, dtype="Int64").to_numpy(float)

~/scipy/pandas/pandas/core/frame.py in to_numpy(self, dtype, copy)
   1335                [2, 4.5, Timestamp('2000-01-02 00:00:00')]], dtype=object)
   1336         """
-> 1337         result = np.array(self.values, dtype=dtype, copy=copy)
   1338         return result
   1339 

TypeError: float() argument must be a string or a number, not 'NAType'
```

So I think that we will need to rework the "DataFrame -> array" conversion a bit (so fix this deeper in the internals). We need to ensure that for EAs, we call `to_numpy` with the kwargs passed through, before combining the different columns in a single array.

Reply:
@jorisvandenbossche Updated to do this in a slightly better way. The below test will fail however as this approach can't guarantee non-copying behavior:
https://github.com/pandas-dev/pandas/blob/085752f95adb625a43aef64481be56a393e8732a/pandas/tests/frame/test_api.py#L363